core/src/Cuda/Kokkos_Cuda_Half.hpp

```diff
@@ -0,0 +1,211 @@
+  return __uint2half_rn(val);
+#else
+  return __float2half(static_cast<float>(val));
+#endif
+}
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(int64_t val) {
+#ifdef __CUDA_ARCH__
+  return __ll2half_rn(val);
+#else
+  return __float2half(static_cast<float>(val));
+#endif
+}
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(uint64_t val) {
+#ifdef __CUDA_ARCH__
+  return __ull2half_rn(val);
+#else
+  return __float2half(static_cast<float>(val));
+#endif
+}
+
+template <class T>
+KOKKOS_INLINE_FUNCTION
+    typename std::enable_if<std::is_same<T, float>::value, T>::type
+    cast_from_half(half_t val) {
+  return __half2float(val);
+}
+template <class T>
+KOKKOS_INLINE_FUNCTION
+    typename std::enable_if<std::is_same<T, double>::value, T>::type
+    cast_from_half(half_t val) {
+  return static_cast<T>(__half2float(val));
+}
+template <class T>
+KOKKOS_INLINE_FUNCTION
+    typename std::enable_if<std::is_same<T, int32_t>::value, T>::type
+    cast_from_half(half_t val) {
+#ifdef __CUDA_ARCH__
+  return __half2int_rz(val);
+#else
+  return static_cast<T>(__half2float(val));
+#endif
+}
+template <class T>
+KOKKOS_INLINE_FUNCTION
+    typename std::enable_if<std::is_same<T, int64_t>::value, T>::type
+    cast_from_half(half_t val) {
+#ifdef __CUDA_ARCH__
+  return __half2ll_rz(val);
+#else
+  return static_cast<T>(__half2float(val));
+#endif
+}
+template <class T>
+KOKKOS_INLINE_FUNCTION
+    typename std::enable_if<std::is_same<T, uint32_t>::value, T>::type
+    cast_from_half(half_t val) {
+#ifdef __CUDA_ARCH__
+  return __half2uint_rz(val);
+#else
+  return static_cast<T>(__half2float(val));
+#endif
+}
+template <class T>
+KOKKOS_INLINE_FUNCTION
+    typename std::enable_if<std::is_same<T, uint64_t>::value, T>::type
+    cast_from_half(half_t val) {
+#ifdef __CUDA_ARCH__
+  return __half2ull_rz(val);
+#else
+  return static_cast<T>(__half2float(val));
+#endif
+}
+
+#else  // CUDA 11.1 below
+
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(float val) { return __float2half(val); }
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(double val) { return __double2half(val); }
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(int32_t val) { return __int2half_rn(val); }
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(uint32_t val) { return __uint2half_rn(val); }
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(int64_t val) { return __ll2half_rn(val); }
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(uint64_t val) { return __ull2half_rn(val); }
+
+template <class T>
+KOKKOS_INLINE_FUNCTION
+    typename std::enable_if<std::is_same<T, float>::value, T>::type
+    cast_from_half(half_t val) {
+  return __half2float(val);
+}
+template <class T>
+KOKKOS_INLINE_FUNCTION
+    typename std::enable_if<std::is_same<T, double>::value, T>::type
+    cast_from_half(half_t val) {
+  return __half2double(val);
+}
+template <class T>
+KOKKOS_INLINE_FUNCTION
+    typename std::enable_if<std::is_same<T, int32_t>::value, T>::type
+    cast_from_half(half_t val) {
+  return __half2int_rz(val);
```

Comment:
Why "round-towards-zero (rz)" rather than 'round-to-nearest-even (rn)" mode?

Reply:
because that is how the default rounding mode of floating point to int works in C and C++ or?